### ISSUE TITLE
should prevent the deletion of a PVC that is referenced by an active pod

### ISSUE THREAD

**Issue by jingxu97** [kubernetes/kubernetes#45143]
The mount/unmount logic is controlled by creation/deletion of the pod. But PV/PVC is controlled by pv controller. As mentioned in issue #44297, in current design, after deleting PVC, the mount point was still exist on the Kubernetes node which the Pod was deployed and also the mount point could be seen from inside the Pod.And also if other Pod reuses the PV, then both old Pod and new Pod could access same mount point.

**Comment by ianchakeres** [comment]
I'm looking at this issue.

**Comment by ianchakeres** [comment]
Here is a method that tests whether a PVC is associated with a PV that is actively in use by a pod.

@jingxu97 - can you please take a look at this code and also let me know if you have any advice on where this method should live?

`

    func isPVCActive(claim *v1.PersistentVolumeClaim)
        // prevent deletion of a pvc that is bound to a pv
        // if the pv is used by an active pod
        // issue #45143

        //check if pvc is bound to pv and associated pv name exists
        if claim != nil && claim.Status.Phase == v1.ClaimBound && claim.Spec.VolumeName != "" {
                //find the associated pv and check
                volume, volumeErr := ctrl.kubeClient.Core().PersistentVolumes().Get(claim.Spec.VolumeName, metav1.GetOptions{})
                if volumeErr == nil {
                        //check all running and pending pods for this volume
                        //podLister is defined outside this method
                        //podLister:  ctrl.kubeClient.Core().Pods().Lister()
                        pods, podErr :=podLister.List(labels.Everything())
                        if podErr == nil {
                                for _, pod := range pods {
                                        if pod.Status.Phase == v1.PodRunning || pod.Status.Phase == v1.PodPending {
                                                for _, podVolume := range pod.Spec.Volumes {
                                                        if podVolume.Name == claim.Spec.VolumeName {
                                                                //found that this pvc uses a pv that is used by an running or pending pod
                                                                ctrl.eventRecorder.Event(claim, v1.EventTypeNormal, "Claim not deleted", "")
                                                                glog.Infof("claim %q:%q not deleted because it is associated with volume %q used by active pod %q:%q", claimToClaimKey(claim), claim.Spec.VolumeName, volume.Name, pod.Namespace, pod.Name)
                                                                return false
                                                        }
                                                }
                                        }
                                }
                        }
                }
        }
        return true
    }

`

**Comment by saad-ali** [comment]
Hey @ianchakeres! Sorry for the delay. Yes that logic looks sound to me. I might change the name from `isPVCActive` to `isPVCOkToDelete`, but that's a nit. My only major concern would be to make sure that by listing all pods we don't introduce any performance regressions, though since it's a one time operation on delete it should be ok.

As for where this code should live? I believe the right place is [plugin/pkg/admission/persistentvolume/label/admission.go](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/persistentvolume/label/admission.go) and doing a check:
```
if a.GetOperation() == admission.Delete && a.GetKind().GroupKind() == api.Kind("PVC")
```

@caesarxuchao, can you verify this is correct: To prevent the deletion of an API object unless certain criteria the check should be added to an admission controller (in this case `admission/persistentvolume`)?

**Comment by caesarxuchao** [comment]
I think implementing in an admission controller is not wrong. Thinking out loud, it would be even better if this could be expressed via [DeleteOptions.Precondition](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L405), though i don't have a concrete idea.

**Comment by saad-ali** [comment]
@lavalamp do you have any sage advice for where best to implement this check to prevent deletion of PVC objects that are still being referenced by a pod?

**Comment by ianchakeres** [comment]
@saad-ali - Thank you for your advice. I've been exploring the location you recommended, and I was able to prevent deletion by replying with an error. I'm actively working to add the related logic to do it more intelligently make a decision based upon the state of the system. If all goes well I'll have a some code for review in the next day or two.

**Comment by saad-ali** [comment]
Ack, thanks @ianchakeres! The admission controller should work, but I'd like to get confirmation from @lavalamp or someone from @kubernetes/sig-api-machinery-misc that it is the best place to add the check. I don't want to send you down the wrong path, so maybe hold off on further work until we get confirmation?

**Comment by deads2k** [comment]
If I'm understanding the flow correctly, the kubelet is using information from the PVC to mount a volume, but is not updating the PVC to reflect a "node X has mounted Y".  If we want a kubelet to indicate that a PVC is "active", why not have a status condition on the PVC that indicates that information?  Doing that would still require some kind of a check server side to prevent removal, but its more likely to be a finalizer that keeps the PVC present until the PVC is inactive so its contained to a single resource and can reflect user intent ("I want this thing gone").

**Comment by saad-ali** [comment]
Thanks for the suggestion @deads2k!

To clarify, the node already keeps track of what volumes are attached or in-use by kubelet via a field in the node status. And attach/detach/mount/unmount logic uses that to prevent premature detaches and unmounts.

The problem is that a pod can reference a PVC/PV pair, which a user may delete before the pod terminates or is deleted. While volume code handles this situation (mostly) gracefully now, it can confuse users (see https://github.com/kubernetes/kubernetes/issues/44297). So the proposal is that if any active pod exists referencing a PVC/PV pair, then we prevent deletion of that PVC until all the pods referencing it are no longer active.

**Comment by smarterclayton** [comment]
How can we transactionally enforce that?  An admission check alone isnt going to allow that behavior.  Without a strong transactional record on the pvc, you really can't ensure that scenario today.

**Comment by smarterclayton** [comment]
I'm tempted to just say "don't do that".  But I understand the desire.

I agree this feels more like a status field on PVC, but it also seems
like this is a UI problem as well ("don't delete this because it's
used by a node" as a message).

Another option is to implement graceful deletion on PVCs and have the
attach controller manage a finalizer (apologies if that's already been
suggested)

**Comment by saad-ali** [comment]
So the proposals for where this logic should exist are:

1. Existing `persistentvolume` admission controller.
2. A separate admission controller that only handles deletes. (@derekwaynecarr's [suggestion](https://github.com/kubernetes/kubernetes/pull/46573#pullrequestreview-41247099)).
3. New `status` field in PVC and a finalizer that keeps the PVC present until the PVC is reported inactive by kubelet. (@deads2k's [suggestion](https://github.com/kubernetes/kubernetes/issues/45143#issuecomment-304270008))
4. Implement graceful deletion on PVCs and have the attach controller manage a finalizer (@smarterclayton's proposal)

I really prefer 1 or 2. While I agree that an admission controller can not guarantee that a volume is never deleted while referenced, since we don't transactionally enforce it (e.g. the admission check can pass, but before the PVC is deleted, a new pod referencing the volume is created). However, it's fairly trivial to implement (https://github.com/kubernetes/kubernetes/pull/46573) and gets us most of the way there. The idea being let's *try* to prevent users from shooting themselves in the foot. If we can do it for the majority of cases (but fail to do so in some race conditions), that should be fine (no need to overengineer).

**Comment by gnufied** [comment]
It would be nice  to have a general solution for this, because this problem keeps coming up in other places too.

Problem with option-3  also is - same PVC can be used in more than one pod. So, if we are going to maintain status field on PVC, we may have to refcount.

**Comment by krmayankk** [comment]
A couple more questions:-
-is admission controller mechanism to delete pvc synchronous ?
-Is there a cheaper method then listing all pods in pvc namespace and compare all volumes in each pod to figure out owning pod ?
-would it be ok to add a label selector to all pv/pvc being referenced by a pod so its easier to find them ?
-would it be ok use ownerreferences (similar to owning controller) to find owning pods for pvc ?

**Comment by saad-ali** [comment]
> -is admission controller mechanism to delete pvc synchronous ?

Nope see comment [above](https://github.com/kubernetes/kubernetes/issues/45143#issuecomment-305361544).

> -Is there a cheaper method then listing all pods in pvc namespace and compare all volumes in each pod to figure out owning pod ?

Maybe maintaining an in-memory cache? Not sure it would be worth optimizing considering infrequency of PVC delete operations.

> -would it be ok to add a label selector to all pv/pvc being referenced by a pod so its easier to find them ?

Would be a pain keeping labels up to date.

> -would it be ok use ownerreferences (similar to owning controller) to find owning pods for pvc ?

2 way pointers are very difficult to manage in a transaction-less system (see PVC/PV).

**Comment by ianchakeres** [comment]
@saad-ali - I prefer option 1. I agree it covers most common case, its easy to understand, and its fairly simple to implement.

**Comment by smarterclayton** [comment]
Before you go much further, can you concretely describe how this is
*not* graceful deletion as we define it today?

Ie how is this different than graceful deletion of pods, namespaces,
or the GC workflow?

**Comment by caesarxuchao** [comment]
It's close to the foreground GC, where an object is deleted only after all its dependents are deleted. The difference is that the garbage collector controller will actively delete the dependents, but here i guess we don't want the GC controller to delete the pods.

IMO the pvc case is different from the graceful deletion of pods in that the [DeletionGracePeriodSeconds](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L179) for the pvc would be infinite.

**Comment by saad-ali** [comment]
> It's close to the foreground GC, where an object is deleted only after all its dependents are deleted. The difference is that the garbage collector controller will actively delete the dependents, but here i guess we don't want the GC controller to delete the pods.
> 
> IMO the pvc case is different from the graceful deletion of pods in that the DeletionGracePeriodSeconds for the pvc would be infinite.

@smarterclayton does that address the graceful deletion/GC approach adequately?

**Comment by smarterclayton** [comment]
Maybe.  Next question - if you go with 1, is that a breaking API change
(before I could write scripts / apply / clients that deleted PVC before
it's done being used)?

**Comment by saad-ali** [comment]
> Maybe.  Next question - if you go with 1, is that a breaking API change (before I could write scripts / apply / clients that deleted PVC before it's done being used)?

In the strictest sense, yes. We have precedence for doing things like this in the past where we think it would be in the user's best interest. If we do go through with this, of course we would have to make sure 1) the new errors are clear and provide enough information on how to recover, 2) broadcast the change loudly in the release notes and discussion lists, and 3) introduce it early enough in 1.8 that it makes it into a few betas and RC.

If we are ok with that, then we can talk about implementation again.

@ianchakeres and @pwittrock had an  offline conversation that I think is worth capturing here:

> Phil - Thank you for making a recommendation and confirming that I'm understanding what you're suggesting. I don't know the answer about what should be performed where. I do know that the admission controllers have access to the informers/listers that are required to make a decision. here's my PR using this approach - https://github.com/kubernetes/kubernetes/pull/46573. The strategy area doesn't have the same plumbing for informers/listers.
> 
> +Saad - since he's helping me get consensus on a design nailed down with apimachinery.
> 
> Saad - Do you think this approach would be better than any of the other paths we're headed in the discussion on issue 45143? https://github.com/kubernetes/kubernetes/issues/45143
> 
> Thanks.
> Ian
> 
> On Thu, Jun 15, 2017 at 8:51 AM, Phillip Wittrock <pwittroc@google.com> wrote:
> Yes, that is generally what I am thinking.  The main question there is whether or not the strategies a intended to perform requests, or are purely static validation.  If doing requests from the strategies is fine, the way you described seems like the way to go.  If not, we should get clear direction on how this sort of resource validation should be done generally.  One thing about the admission controllers is that many of them are optional, and can be turned on or off.  This seems like something that should be built directly into the PVC api logic itself, instead of as an add-on component.
> 
> FYI, This is outside my direct area of ownership, and you will need to get input from sig-apimachinery and their blessing if we are going to update the Strategy.
> 
> I would start by heading to the sig-apimachinery slack channel and getting initial feedback for the idea there.  Be sure to highlight that you would need to do lookups from the strategy.  Then file either an issue or design proposal if there is some general support for the idea.
> 
> If you are eager to get started, you could implement the proof of concept code after sending out the proposal, but do so knowing that the proposal maybe rejected in favor of some alternative, or deferred due to code review resource constraints.
>
>
>
> On Thu, Jun 15, 2017 at 6:33 AM Ian Chakeres <ichakeres@salesforce.com> wrote:
> Hi Phil,
>
> I investigated your recommendation to update DeleteStrategy, and I want to make sure that I’m headed in the right direction.
> 
> We would add a function to the RESTDeleteStrategy Interface, e.g. CheckDelete (similar to CheckGracefulDelete). Inside BeforeDelete we’d call the CheckDelete function. We’d modify the persistent volume claim strategy to implement the CheckDelete function. Inside CheckDelete we’d put the logic about deletion for PVCs and whether they can be safely deleted.
> 
> This implementation would enable other objects to handle a similar situation in the future in a more generic fashion. 
> 
> Am I barking up the right tree?
> 
> If you think that a real-time chat might be better than email and have 10m, feel free to give me a ring during your commute.
> 
> Thanks.
>Ian 
> 
> On Tue, Jun 13, 2017 at 9:08 AM, Phillip Wittrock <pwittroc@google.com> wrote:
> <snip>
> 
> #3 preventing deletion of PVC used by active pod. Here’s the discussion - https://github.com/kubernetes/kubernetes/issues/45143. Here’s my PR - https://github.com/kubernetes/kubernetes/pull/46573. You could chime in on the thread or just tell me where/how you think this logic should be implemented. Any advice and examples would be much appreciated.
> 
> I haven't dug into this deeply, but my gut says this is the wrong direction, and we should instead update the DeleteStrategy to have an option for whether an object is ok to delete.  1) Imagine other resources may have similar requirements, and this will provide better support for that.  2) The customized logic for resource CRUD operations shouldn't be spread across the code base, but should instead live within the storage strategies.  If this is urgent, the AC may be an ok hack to get it done, but it is definitely technical debt.
>
> 
> Is there anything I can do to help you out?
> 
> Have a great week.
> Ian 

My 2 cents: 

>  I think extending DeleteStrategy to support this use case would be better than using an Admission Controller. The argument that rings true for me is: "One thing about the admission controllers is that many of them are optional, and can be turned on or off.  This seems like something that should be built directly into the PVC api logic itself, instead of as an add-on component."

**Comment by gnufied** [comment]
Can we initialize informers in strategies? This is something, that  was a big no-no, last I looked. cc @liggitt

**Comment by smarterclayton** [comment]
No, apiserver is stateless and shouldn't assume global visibility.
Admission control can judiciously use caches but it comes at a cost.
I'd prefer you didn't have to cache all pods to calculate use,
although we could maybe add a field filter if we had to

I'm pretty sure the right behavior here is a finalizer and graceful
deletion given the design of the overall system.  I'm ok with an
admission controller stopgap, although I do not think it is the right
thing for the  long term.  Attach detach controller could handle this
responsibility fairly effectively today, and should have all the
necessary caches in place.  I guess I'm not sure why an admission
controller is easier, but I won't block you on this if you think it
is.

**Comment by ianchakeres** [comment]
On Thursday (6/22/2017) I met with @lavalamp to chat about the various options discussed in this thread. CC: @saad-ali 

We focused on two options that could be used to resolve this issue:
1. Bind PVC and pods more closely using multi-phase locking
   * PVCs maintain references to related pods when in particular states. Related PVs should not be used until the pod/pvc relationship is committed. When someone tries to delete a PVC, these references would be checked and an error immediately returned if a related PVC-PVs is actively used.

2. Use a PVC finalizer
    * PVCs would have a finalizer. During deletion this finalizer determines whether the PVC can be deleted. If it cannot be deleted, the PVC will remain pending deletion and periodically re-checked. The pending delete state will prevent the PVC-PV from being associated/mounted. 

Between these two options, I prefer the finalizer approach. This approach has looser coupling between objects. I also like that it eventually deletes the PVC, achieving the desired state from an operator. It could take a while for the deletion to complete, and during this phase the operator can observe the state of the PVC.

I’d like to hear from others whether they have a preference between these two options and why they prefer one option or another. Feel free to describe or re-energize other options too.

**Comment by smarterclayton** [comment]
The first one doesn't scale well, the latter sounds like a perfect
match for the finalizer usecase (but you need to mark the pvc as being
in graceful deletion to prevent new pods and kubelets from being able
to use it).

**Comment by gnufied** [comment]
There is a plan to use attach/detach controller as a fencing mechanism to prevent RWX of RWO volume types. As part of that effort, I am proposing that - we introduce a `node` field in PVC status (for now I am considering PVC but it can change to PV). 

This field is updated on attach and detach. All volume types implement either a logical attach/detach(iSCSI, NFS etc) or real attach/detach and hence `status.node` field gets updated accordingly.  If we have such a field on PVC, can we not simply stop PVC deletion via validation?

cc @rootfs

**Comment by smarterclayton** [comment]
That's another option.  Certainly RWM doesn't need this construct if so.

**Comment by thockin** [comment]
In my mind, this is what finalizers are good for.  We don't often use them as such, especially because transactions don't exist, but this is what they are for.

**Comment by jsafrane** [comment]
Resuming an old thread..

> There is a plan to use attach/detach controller as a fencing mechanism to prevent RWX of RWO volume types

A/D controller is not a good place for this. It can prevent from one RWO PV to be attached to multiple nodes but it has no way how to prevent from two pods using the same RWO PV on the same node. These two pods expect RWO and can corrupt data of each other.

**Comment by jsafrane** [comment]
Regarding finalizers, last time I checked they could be used to delay deletion of something, but they could not prevent it. IMO, when I delete an object with a finalizer, the object gets a grace period and it will be forcefully deleted after this period. The message to user is "hold on, we're working on deleting the object". So a PVC with a finalizer will be deleted even if there is a pod that uses it.

To implement real finalizers, we would need to do the cascade delete - if a PVC is deleted then delete all pods that use it. If PV is deleted then mark the PVC as lost and delete all the pods too.

**Comment by pospispa** [comment]
Please, let me re-phrase the bug description using Attacker and Victim wording.

1. Admin creates RWX PVs that are ready to be bound to newly created PVCs.
2. Attacker creates PVC1 that bounds to PV1.
3. Attacker creates a pod that uses the PVC1 so the PV1 is mounted to the attacker's pod.
4. The attacker's pod keeps running but the application in the pod is not actively using the mounted PV1.
5. The attacker deletes the PVC1.
6. Victim creates PVC2 that is bound to the PV1 that is mounted by the attacker's pod.
7. Victim creates a pod that uses the PVC2 so the PV1 is now mounted to both attacker's and victim's pod.
8. Victim's pod writes data into PV1 and these data are immediately available to the attacker's pod.

I think this is a high-priority security bug.

Please, do you think that this bug is a release blocker?

I don't have rights to update the issue labels so I'll appreciate if whoever has the rights will update the labels.

**Comment by liggitt** [comment]
> Regarding finalizers, last time I checked they could be used to delay deletion of something, but they could not prevent it. IMO, when I delete an object with a finalizer, the object gets a grace period and it will be forcefully deleted after this period. The message to user is "hold on, we're working on deleting the object". So a PVC with a finalizer will be deleted even if there is a pod that uses it.

That is not correct, finalizers block deletion.

https://github.com/kubernetes/kubernetes/blob/b188868fd929593a0fcf41a7ee2afe3e10b03c39/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L886-L889

However, nothing prevents a user with write permissions on a PVC from removing the finalizer. @caesarxuchao, this is why we needed to enforce addition/removal of finalizers.

> To implement real finalizers, we would need to do the cascade delete - if a PVC is deleted then delete all pods that use it. If PV is deleted then mark the PVC as lost and delete all the pods too.

No, a normal finalizer should be sufficient. Separately, we should require finalizers permission on the object to add/remove from that list.

> I think this is a high-priority security bug.

Reuse of PVs is less than ideal, in both directions:
* ensuring data is scrubbed so the old user's data is guaranteed not to be accessible to the new user is difficult
* ensuring existing uses are removed is not done today (this issue)

This is only an issue with a reclaim policy of recycle, correct? For environments where strict PV tenancy is required, a reclaim policy of Delete or Retain would prevent rebinding the PV to another PVC.

> Please, do you think that this bug is a release blocker?

No, since it's not a regression or change in behavior from a previous release, and reclaim policy can be used to control rebind behavior.

**Comment by jsafrane** [comment]
> That is not correct, finalizers block deletion.

Ack, they're hidden behind `kubectl get --show-all`.

Still, isn't it a bad UX? Kubectl should display an error that PVC can't be deleted because of something uses it. How is user supposed to know why it's not deleted?

**Comment by liggitt** [comment]
issuing a delete call, setting a deletionTimestamp and persisting, *is* successful deletion initiation. the UX of things pending deletion could be improved, sure.

**Comment by jsafrane** [comment]
[Namespace finalizer owners](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/namespaces.md#finalizers) start actively working on deleting stuff in a namespace that's being deleted. User knows it's going to be deleted eventually and it's rather quick in most cases. User can see progress and check what's blocking the deletion. That's what I would call "deletion initiation".

Nothing happens when someone deletes a PVC with a finalizer. It's just persisted. Maybe forever. Should some controller start deleting the pods? It's not hard and IMO it would solve a real bug.

**Comment by liggitt** [comment]
> Nothing happens when someone deletes a PVC with a finalizer. It's just persisted. Maybe forever. Should some controller start deleting the pods? It's not hard and IMO it would solve a real bug.

Yes, that's how finalizers work. Every finalizer listed in an object's metadata.finalizer should have an associated controller that is responsible for finalizing and then removing that finalizer name. That is why a normal user should not be able to add/remove things from that field.

**Comment by gnufied** [comment]
> Every finalizer listed in an object's metadata.finalizer should have an associated controller that is responsible for finalizing 

Does that have to be some special controller? or any controller (say attach/detach) can watch for pod deletion events and remove finalizer from associated PVC objects?

**Comment by liggitt** [comment]
> Does that have to be some special controller? or any controller (say attach/detach) can watch for pod deletion events and remove finalizer from associated PVC objects?

It should be a controller built specifically for that purpose. You could go a couple directions:
* wait until pods referenced by a pending-finalization PVC terminate on their own, then remove the finalizer from the PVC
* proactively delete pods that reference a pending-finalization PVC, and once they are deleted, remove the finalizer from the PVC

**Comment by thockin** [comment]
I don't think we want to do cascading delete.  At the VERY LEAST, we do not
want to do it ad hoc for one case.  We have no other formalized
referential-integrity mechanisms.  If we want to consider such semantics,
it would effectively be an overhaul of the API and all of the associated
machinery.

Finalizers allow the DELETE op to start, but not finish.  It's not exactly
delete inhibition, which is what we want, but it's close.  I don't think we
want a parallel list of delete inhibitors, but that is an API machinery
question.

**Comment by pospispa** [comment]
IMHO, the discussion came to conclusion that the approach with the smallest amount of disadvantages is to use finalizer to postpone the PVC deletion in case it's used by a pod.

I've written [design](https://github.com/kubernetes/community/pull/1174) for using the finalizer. I hope it's not an overkill to write a design for fixing this issue.
